DirectProgramming/DPC++FPGA/Tutorials/Features/floating_point_pragmas/src/fp_pragmas.cpp

```diff
@@ -0,0 +1,269 @@
+  try {
+    queue q(selector, dpc_common::exception_handler,
+            property::queue::enable_profiling{});
+
+    buffer<double, 1> bufferA(&a, 1);
+    buffer<double, 1> bufferB(&b, 1);
+    buffer<double, 1> bufferC(&c, 1);
+    buffer<double, 1> bufferD(&d, 1);
+    buffer<double, 1> bufferE(&res, 1);
+
+    event e = q.submit([&](handler &h) {
+      accessor accessorA(bufferA, h, read_only);
+      accessor accessorB(bufferB, h, read_only);
+      accessor accessorC(bufferC, h, read_only);
+      accessor accessorD(bufferD, h, read_only);
+      accessor accessorRes(bufferE, h, write_only, noinit);
+
+      // The "kernel_args_restrict" tells the compiler that a, b, and r
+      // do not alias. For a full explanation, see:
+      //    DPC++FPGA/Tutorials/Features/kernel_args_restrict
+      h.single_task<ReassociateOffKernel>([=
+      ]() [[intel::kernel_args_restrict]] {
+        accessorRes[0] =
+            accessorA[0] + accessorB[0] + accessorC[0] + accessorD[0];
+      });
+    });
+  } catch (sycl::exception const &e) {
+    // Catches exceptions in the host code
+    std::cerr << "Caught a SYCL host exception:\n" << e.what() << "\n";
+
+    // Most likely the runtime couldn't find FPGA hardware!
+    if (e.get_cl_code() == CL_DEVICE_NOT_FOUND) {
+      std::cerr << "If you are targeting an FPGA, please ensure that your "
+                   "system has a correctly configured FPGA board.\n";
+      std::cerr << "Run sys_check in the oneAPI root directory to verify.\n";
+      std::cerr << "If you are targeting the FPGA emulator, compile with "
+                   "-DFPGA_EMULATOR.\n";
+    }
+    std::terminate();
+  }
+}
+
+// Launch a kernel on the device specified by selector.
+// The kernel's functionality is designed to show the
+// performance impact of the fp contract pragma.
+void enable_contract(const device_selector &selector, const double a,
+                     const double b, const double c, const double d,
+                     double &res) {
+  try {
+    queue q(selector, dpc_common::exception_handler,
+            property::queue::enable_profiling{});
+
+    buffer<double, 1> bufferA(&a, 1);
+    buffer<double, 1> bufferB(&b, 1);
+    buffer<double, 1> bufferC(&c, 1);
+    buffer<double, 1> bufferD(&d, 1);
+    buffer<double, 1> bufferE(&res, 1);
+
+    event e = q.submit([&](handler &h) {
+      accessor accessorA(bufferA, h, read_only);
+      accessor accessorB(bufferB, h, read_only);
+      accessor accessorC(bufferC, h, read_only);
+      accessor accessorD(bufferD, h, read_only);
+      accessor accessorRes(bufferE, h, write_only, noinit);
+
+      // The "kernel_args_restrict" tells the compiler that a, b, and r
+      // do not alias. For a full explanation, see:
+      //    DPC++FPGA/Tutorials/Features/kernel_args_restrict
+      h.single_task<ContractFastKernel>([=]() [[intel::kernel_args_restrict]] {
+
+      // Allows the fusing of mult and add instructions into an FMA.
```

Comment:
How about:

// This pragma allows re-ordering of the multiply and add instructions to an order that maps more efficiently into FPGA hardware